solr/solrj/src/java/org/apache/solr/client/solrj/cloud/DirectShardState.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.client.solrj.cloud;
+
+import java.util.function.Predicate;
+
+import org.apache.solr.common.cloud.Replica;
+import org.apache.solr.common.cloud.Slice;
+
+/**Reads the state from state.json
+ *
+ */
+public class DirectShardState implements ShardStateProvider {
```

Comment:
IMHO having this class is harmful to the intended goal, because it maintains the illusion that replica state is actually kept up to date inside the `Replica` object. The whole point of this refactoring, as I understand it, is that the state is maintained elsewhere.
I would even go as far as to completely remove `Replica.getState()`.

Reply:
How else can we achieve the goal. The idea is for both of these to co-exist for a few releases. Eventually, this can go away

Reply:
I think we have to decide where we keep (and maintain) the actual state. The API that you proposed provides two ways of getting this info, and two potential locations to get it (which may get out of sync, or have different staleness), which in my opinion is one too many :)

If the replica state resides in `Replica` and this is our source of truth then having a `ShardStateProvider` seems pointless. What is the source of truth then - the state that is cached in `Replica` or the state that we obtain from `ShardStateProvider` implementation, which is guaranteed to the the most up-to-date?

My understanding was that the purpose of introducing `ShardStateProvider` was to provide The single API point that is the only source of truth about the current shard/replica state (similar to other *StateProvider-s). In this model `Replica` becomes a simple static descriptor and it's NOT responsible for caching its state - that's the role of `ShardStateProvider`.

Reply:
>My understanding was that the purpose of introducing ShardStateProvider was to provide The single API point that is the only source of truth about the current shard/replica state (similar to other *StateProvider-s)

The `state` attribute will have to go away when the old model goes away. 

So, if your collection is created with `externalState=true` You really have no information of the state in the `state.json`. Nobody updates it nobody reads from it. Internally we always must rely on the interface to do the right thing.